20220831-add-type-restrictions-to-json-syntax.md

```diff
@@ -0,0 +1,461 @@
+
+# Definitions
+
+- [OpenFGA JSON Syntax](https://openfga.dev/docs/configuration-language)
+- [OpenFGA DSL](https://openfga.dev/docs/configuration-language)
+- [ListObjects Endpoint](https://openfga.dev/api/service#/Relationship%20Queries/ListObjects)
+
+# Motivation
+
+## Why should we do this?
+
+### Optimize ListObjects
+In order to implement some optimizations to the ListObjects endpoint mentioned [ListObjects RFC](https://github.com/openfga/rfcs/blob/main/20220714-listObjects-api.md), we should be able to traverse the relationship graph in reverse.
+
+### Improve the Developer Experience
+At the moment, our authorization model does not allow users to indicate what the type they expect on a relation to be, leaving it a bit open-ended and up for interpretation, especially when someone is reading the authorization model later without much context and where the original authors are not available. It is hard for someone to grok from first glance that a repository owner should be an organization or a user. Adding type restrictions allows developers to have an easier time in interpreting the model.
+
+## What is the expected outcome?
+
+- Using the extra type information we get from this change, we will be able to more easily traverse the graph in reverse, thus allowing us to provide a solution for reverse expansion and optimizing the ListObjects endpoint.
+- Users reading an authorization model will be able to better interpret its meaning (for example by understanding that repository owners must be either users or group members).
+
+# What it is?
+This RFC proposes an update to the OpenFGA JSON syntax requiring users to indicate all the types of users that could be **directly related** to an object of a certain type through a particular relation.
+
+For example,
+- parents of a document have to be objects of type folder
+- owners of a repository have to be of type user or a userset of group members
+
+We will add an `assignable_relations` to each relation to indicate what types of users can be directly related to it.
+
+`assignable_relations` will be an array of objects, each object must have a type and an optional relation.
+- having the following `"parent": { "assignable_relations": [{ "type": "group" }] }` in the group type definition indicates that only users of type group can be directly related to a group as parent
+- having the following `"member": { "assignable_relations": [{ "type": "user" }, { "type": "group", "relation": "member" }] }` in the group type definition indicates that only objects of type user or usersets of type group and relation member can be directly related to a group as member
+- having the following `"member": { "assignable_relations": [{ "type": "user:*" }, { "type": "employee:*" }] }` in the group type definition indicates that the `*` syntax is allowed and that when present, it means all objects of type user or group can be directly related to a group as member
```

Comment:
I am confused with the sentence.  Where does `group` come into the definition? I only see `user` and `employee` in the model.  Also, maybe more explicit in saying that `*` is not allowed in the other two cases?

Reply:
`member` is a relation of `group`.